src/python/pants/engine/target.py

```diff
@@ -714,6 +714,52 @@ def expect_single(self) -> Target:
         return self[0]
 
 
+@dataclass(frozen=True)
+class ShouldTraverseDepsPredicate(metaclass=ABCMeta):
+    """This callable determines whether to traverse through deps of a given Target + Field.
+
+    This is a callable dataclass instead of a function to avoid issues with hashing closures.
+    Only the id is used when hashing a function; any closure vars are NOT accounted for.
+
+    NB: When subclassing a dataclass (like this one), generated methods on a dataclass are NOT
+    inherited because the @dataclass decorator only inspects the current class before generating
+    a function--the decorator does NOT inspect __mro__.
+    """
+
+    # This _callable field ensures that __call__ is included in the __hash__ method generated by @dataclass.
```

Comment:
```suggestion
    # NB: This _callable field ensures that __call__ is included in the __hash__ method generated by @dataclass.
    #  That is extremely important because two predicates with different implementations but the same data (or no data)
    # need to have different hashes and compare unequal.
```